File: backend/control.py
"""Wrapper komunikasi serial dengan Arduino (scanning fotoakustik)."""

import threading
import time

import serial
import serial.tools.list_ports


class SerialController:
    def __init__(self, on_message=None, on_status_change=None):
        self.ser = None
        self.read_thread = None
        self.running = False
        self.on_message = on_message
        self.on_status_change = on_status_change

    @staticmethod
    def list_ports():
        hasil = []
        for p in serial.tools.list_ports.comports():
            deskripsi = p.description if p.description and p.description != "n/a" else ""
            label = f"{p.device} - {deskripsi}" if deskripsi else p.device
            hasil.append((p.device, label))
        return hasil

    def connect(self, port, baudrate=9600, timeout=1):
        try:
            self.ser = serial.Serial(port, baudrate, timeout=timeout)
            time.sleep(2)
            self.running = True
            self.read_thread = threading.Thread(target=self._read_loop, daemon=True)
            self.read_thread.start()
            if self.on_status_change:
                self.on_status_change(True)
            return True, f"Terhubung ke {port} @ {baudrate} baud"
        except Exception as e:
            self.ser = None
            return False, str(e)

    def disconnect(self):
        self.running = False
        if self.read_thread is not None:
            self.read_thread.join(timeout=1)
        if self.ser is not None and self.ser.is_open:
            self.ser.close()
        self.ser = None
        if self.on_status_change:
            self.on_status_change(False)

    def is_connected(self):
        return self.ser is not None and self.ser.is_open

    def send(self, command):
        if not self.is_connected():
            return False, "Belum terhubung ke Arduino"
        try:
            self.ser.write((command.strip() + "\n").encode("utf-8"))
            return True, f"Terkirim: {command}"
        except Exception as e:
            return False, str(e)

    def set_x(self, cm):
        return self.send(f"x={cm}")

    def set_y(self, cm):
        return self.send(f"y={cm}")

    def start_scan(self):
        return self.send("start")

    def stop_scan(self):
        return self.send("stop")

    def jog_kanan(self):
        return self.send("kanan")

    def jog_kiri(self):
        return self.send("kiri")

    def jog_maju(self):
        return self.send("maju")

    def jog_mundur(self):
        return self.send("mundur")

    def set_laser_freq(self, hz):
        """Kirim f=<Hz> ke Arduino 1; diteruskan ke Arduino laser via SoftSerial."""
        try:
            nilai = float(hz)
        except (TypeError, ValueError):
            return False, "Frekuensi laser tidak valid"
        if nilai < 0.1 or nilai > 50000.0:
            return False, "Frekuensi laser di luar rentang (0.1 .. 50000 Hz)"
        if float(nilai).is_integer():
            teks = str(int(nilai))
        else:
            teks = f"{nilai:.2f}"
        return self.send(f"f={teks}")

    def _read_loop(self):
        while self.running and self.ser is not None and self.ser.is_open:
            try:
                raw = self.ser.readline()
                if not raw:
                    continue
                line = raw.decode("utf-8", errors="replace").strip()
                if line and self.on_message:
                    self.on_message(line)
            except Exception as e:
                if self.on_message:
                    self.on_message(f"[ERROR BACA SERIAL] {e}")
                break

File: cpp/backend/control.cpp
#include "backend/control.hpp"

#include <cmath>

#include <QSerialPortInfo>
#include <QThread>

#include "backend/config.hpp"

namespace pa {

SerialController::SerialController(QObject* parent) : QObject(parent) {}

SerialController::~SerialController() { disconnectPort(); }

std::vector<std::pair<QString, QString>> SerialController::listPorts() {
  std::vector<std::pair<QString, QString>> out;
  for (const QSerialPortInfo& info : QSerialPortInfo::availablePorts()) {
    const QString device = info.portName();
    QString label = device;
    if (!info.description().isEmpty())
      label = device + " - " + info.description();
    out.emplace_back(device, label);
  }
  return out;
}

std::pair<bool, QString> SerialController::connectTo(const QString& port,
                                                     int baudrate) {
  disconnectPort();
  port_ = new QSerialPort(this);
  port_->setPortName(port);
  port_->setBaudRate(baudrate);
  port_->setDataBits(QSerialPort::Data8);
  port_->setParity(QSerialPort::NoParity);
  port_->setStopBits(QSerialPort::OneStop);
  port_->setFlowControl(QSerialPort::NoFlowControl);
  if (!port_->open(QIODevice::ReadWrite)) {
    const QString err = port_->errorString();
    delete port_;
    port_ = nullptr;
    return {false, err};
  }
  // Arduino reset delay (mirip Python time.sleep(2))
  QThread::msleep(2000);
  connect(port_, &QSerialPort::readyRead, this, [this]() {
    while (port_ && port_->canReadLine()) {
      const QByteArray raw = port_->readLine();
      const QString line = QString::fromUtf8(raw).trimmed();
      if (!line.isEmpty()) emit messageReceived(line);
    }
  });
  emit statusChanged(true);
  return {true, QString("Terhubung ke %1 @ %2 baud").arg(port).arg(baudrate)};
}

void SerialController::disconnectPort() {
  if (!port_) return;
  port_->close();
  port_->deleteLater();
  port_ = nullptr;
  emit statusChanged(false);
}

bool SerialController::isConnected() const {
  return port_ && port_->isOpen();
}

std::pair<bool, QString> SerialController::send(const QString& command) {
  if (!isConnected()) return {false, "Belum terhubung ke Arduino"};
  const QByteArray data = (command.trimmed() + "\n").toUtf8();
  const qint64 n = port_->write(data);
  if (n < 0) return {false, port_->errorString()};
  port_->flush();
  return {true, QString("Terkirim: %1").arg(command)};
}

std::pair<bool, QString> SerialController::setX(double cm) {
  return send(QString("x=%1").arg(cm));
}
std::pair<bool, QString> SerialController::setY(double cm) {
  return send(QString("y=%1").arg(cm));
}
std::pair<bool, QString> SerialController::startScan() { return send("start"); }
std::pair<bool, QString> SerialController::stopScan() { return send("stop"); }
std::pair<bool, QString> SerialController::jogKanan() { return send("kanan"); }
std::pair<bool, QString> SerialController::jogKiri() { return send("kiri"); }
std::pair<bool, QString> SerialController::jogMaju() { return send("maju"); }
std::pair<bool, QString> SerialController::jogMundur() { return send("mundur"); }

std::pair<bool, QString> SerialController::setLaserFreq(double hz) {
  if (hz < 0.1 || hz > 20000.0)
    return {false, "Frekuensi laser di luar rentang (0.1 .. 20000 Hz)"};
  QString teks;
  if (std::abs(hz - std::round(hz)) < 1e-9)
    teks = QString::number(static_cast<int>(std::lround(hz)));
  else
    teks = QString::number(hz, 'f', 2);
  return send(QString("f=%1").arg(teks));
}

}  // namespace pa
